File: src/main/java/com/motio/model/Meal.java
package com.motio.model;

import jakarta.persistence.*;
import jakarta.validation.constraints.NotBlank;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.util.HashSet;
import java.util.LinkedList;
import java.util.List;
import java.util.Set;

@Data
@NoArgsConstructor
@AllArgsConstructor
@Entity
@Table(name = "meals")
public class Meal {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @NotBlank(message = "Meal name is mandatory")
    @Column(nullable = false)
    private String mealName;

    @ManyToOne
    @JoinColumn(name = "created_by_user_id", nullable = false)
    private User createdByUser;

    @ManyToMany
    @JoinTable(
            name = "meal_user_access",
            joinColumns = @JoinColumn(name = "meal_id"),
            inverseJoinColumns = @JoinColumn(name = "user_id")
    )
    private Set<User> accessibleUsers = new HashSet<>();

    @ManyToMany
    @JoinTable(
            name = "meal_and_meal_category_association",
            joinColumns = @JoinColumn(name = "meal_id"),
            inverseJoinColumns = @JoinColumn(name = "meal_category_id")
    )
    private Set<MealCategory> categories;

    @ElementCollection
    @CollectionTable(name = "meal_steps", joinColumns = @JoinColumn(name = "meal_id"))
    @Column(name = "step")
    private List<String> steps = new LinkedList<>();

    @ElementCollection
    @CollectionTable(name = "meal_ingredients", joinColumns = @JoinColumn(name = "meal_id"))
    @Column(name = "ingredient")
    private List<String> ingredients = new LinkedList<>();

    @Column
    private String imagePath;
}

File: motio_client/lib/models/meal.dart
import 'package:motio_client/models/user.dart';

import 'meal_category.dart';

class Meal {
  int id;
  String mealName;
  User createdByUser;
  Set<User> accessibleUsers;
  Set<MealCategory> categories;
  List<String> steps;
  List<String> ingredients;
  String? imagePath; // Nullable field

  Meal({
    required this.id,
    required this.mealName,
    required this.createdByUser,
    required this.accessibleUsers,
    required this.categories,
    required this.steps,
    required this.ingredients,
    this.imagePath, // Nullable field
  });

  factory Meal.fromJson(Map<String, dynamic> json) {
    return Meal(
      id: json['id'],
      mealName: json['mealName'],
      createdByUser: User.fromJson(json['createdByUser']),
      accessibleUsers: (json['accessibleUsers'] as List)
          .map((userJson) => User.fromJson(userJson))
          .toSet(),
      categories: (json['categories'] as List)
          .map((categoryJson) => MealCategory.fromJson(categoryJson))
          .toSet(),
      steps: List<String>.from(json['steps']),
      ingredients: List<String>.from(json['ingredients']),
      imagePath: json['imagePath'], // Nullable field
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'mealName': mealName,
      'createdByUser': createdByUser.toJson(),
      'accessibleUsers': accessibleUsers.map((user) => user.toJson()).toList(),
      'categories': categories.map((category) => category.toJson()).toList(),
      'steps': steps,
      'ingredients': ingredients,
      'imagePath': imagePath, // Nullable field
    };
  }
}
